keras_cv/layers/vit_layers_test.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv.layers.vit_layers import PatchEmbedding
+from keras_cv.layers.vit_layers import Patching
+
+
+class ViTLayersTest(tf.test.TestCase):
+    def test_patching_return_type_and_shape(self):
+        layer = Patching(patch_size=16)
+
+        inputs = tf.random.normal([1, 224, 224, 3])
+        output = layer(inputs)
+        self.assertTrue(isinstance(output, tf.Tensor))
+        self.assertLen(output, 1)
+        self.assertEquals(output.shape, [1, 196, 768])
+
+    def test_patching_wrong_patch_size(self):
+        layer = Patching(patch_size=-16)
+        inputs = tf.random.normal([1, 224, 224, 3])
+        with self.assertRaisesRegexp(
+            ValueError,
+            "The patch_size cannot be a negative number. Received -16",
+        ):
+            layer(inputs)
+
+    def test_patching_wrong_padding(self):
+        with self.assertRaisesRegexp(
+            ValueError,
+            "Padding must be either 'SAME' or 'VALID', but REFLECT was passed.",
+        ):
+            Patching(patch_size=16, padding="REFLECT")
+
+    def test_patch_embedding_return_type_and_shape(self):
+        layer = PatchEmbedding(project_dim=128)
+        inputs = tf.random.normal([1, 196, 768])
+        output = layer(inputs)
+        self.assertTrue(isinstance(output, tf.Tensor))
+        self.assertLen(output, 1)
+        self.assertEquals(output.shape, [1, 197, 128])
+
+    def test_patch_embedding_interpolation(self):
+        inputs = tf.random.normal([1, 196, 768])
+        layer = PatchEmbedding(project_dim=128)
+
+        output = layer(
+            patch=inputs,
+            interpolate=True,
+            interpolate_width=300,
+            interpolate_height=300,
+            patch_size=12,
+        )
+
+        self.assertTrue(isinstance(output, tf.Tensor))
```

Comment:
Also add some numerical tests if possible -- especially on the interpolation code path

Reply:
Possibly a stupid question - but can we enforce numerical tests here? There are multiple random elements in the code, so we'd get different results on each run, even with the same input and even when setting a global seed with `tf.keras.utils.set_random_seed(seed)`.

I figured we could probably test the input embeddings against the interpolated and then sampled down embeddings (because they should be approximately the same) but the input to the layer is a sequence of patches, and the result is a linear projection which no longer resembles images but an arbitrary learnable vector, so that numerical test falls down the drain :/

Open to ideas!

Reply:
we should test the numerical outputs of the patching logic, if possible

Reply:
The interpolated output is also summed with a Dense projection before being returned:

```
reshaped_embeddings  = #... resizing, reshaping, etc. with interpolation
linear_projection = self.linear_projection(reshaped_embeddings)
addition = linear_projection + reshaped_embeddings

return tf.concat([class_token, addition], 1)
```

The `Dense` call here that's added to the `reshaped_embeddings` will necessarily add randomness to the `__interpolate_positional_embedings` output. I'll separate the logic for `__interpolate_positional_embeddings` to only interpolate and return the embeddings, and then do the linear projection and summation outside the method. This should make the method testable...

For the other one - sounds good! I'll add that in :)

Reply:
Done! I changed the logic for the interpolation method to just interpolate and the addition with the linear projection is now done outside of it. With this, we can check whether an all 1s tensor is still interpolated as all 1s in the end. The patching layer was tested on this image:

<img width="287" alt="image" src="https://user-images.githubusercontent.com/60978046/201557006-c3027275-6fa4-4145-86c8-0b28e4c38b2f.png">

And as to whether it correctly returns:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/60978046/201557031-c5704e79-2b4a-45d3-a8c1-d713172a0386.png">

...numerically. Thanks for requesting these! Could you take a look? @tanzhenyu